pubsub/azurepubsub/azurepubsub.go

```diff
@@ -46,245 +49,161 @@ const (
 }
 
 // SubscriptionOptions will contain configuration for subscriptions.
 type SubscriptionOptions struct{
 	ListenerTimeout time.Duration
 }
 
-// OpenSubscription opens a ServiceBus Subscription on the given topicName and namespace connection string.
-func OpenSubscription(ctx context.Context, topicName string, subscriptionName string, connString string, opts *SubscriptionOptions) (*pubsub.Subscription, error) {
-	s, err := openSubscription(ctx, topicName, subscriptionName, connString, opts)
+// OpenSubscription initializes a pubsub Subscription on a given Service Bus Subscription and its parent Service Bus Topic.
+func OpenSubscription(ctx context.Context, sbSubscription *servicebus.Subscription, parentTopic *servicebus.Topic, parentNamespace *servicebus.Namespace, opts *SubscriptionOptions) (*pubsub.Subscription, error) {
+	s, err := openSubscription(ctx, sbSubscription, parentTopic, parentNamespace, opts)
 	if err != nil {
 		return nil, err
 	}
-
 	return pubsub.NewSubscription(s, nil), nil
 }
 
 // openSubscription returns a driver.Subscription.
-func openSubscription(ctx context.Context, topicName string, subscriptionName string, connectionString string, opts *SubscriptionOptions) (driver.Subscription, error) {
-	ns, err := getSbNamespace(connectionString)
-	if err != nil{
-		return nil, err
+func openSubscription(ctx context.Context, sbSub *servicebus.Subscription, sbTop *servicebus.Topic, sbNs *servicebus.Namespace, opts *SubscriptionOptions) (driver.Subscription, error) {	
+	if sbTop == nil {	
+		return nil, errors.New("azurepubsub: cannot initialize driver.Subscription with a nil Service Bus Topic")
+	}
+	if sbNs == nil {		
+		return nil, errors.New("azurepubsub: cannot initialize driver.Subscription with a nil Service Bus Namespace")
+	}
+
+	if sbSub == nil {		
+		return nil, errors.New("azurepubsub: cannot initialize driver.Subscription with a nil Service Bus Subscription")
 	}
 
 	defaultTimeout := listenerTimeout
-	if opts != nil {
-		if opts.ListenerTimeout > 0 {
-			defaultTimeout = opts.ListenerTimeout
-		}
+	if opts != nil && opts.ListenerTimeout > 0 {
+		defaultTimeout = opts.ListenerTimeout		
 	}
 
 	return &subscription {		
-		name : subscriptionName,
-		topicName: topicName,		
-		ns: ns,
+		sbSub : sbSub,
+		topicName: sbTop.Name,		
+		sbNs: sbNs,
 		opts : &SubscriptionOptions {
 			ListenerTimeout: defaultTimeout,
 		},
 	}, nil
 }
 
-func createSubscription(ctx context.Context, topicName string, subscriptionName string, connectionString string, opts[] servicebus.SubscriptionManagementOption) (error) {
-	ns, err := getSbNamespace(connectionString)
-	if err != nil {
-		return err
-	}
-
-	subManager, err := ns.NewSubscriptionManager(topicName)
+// testSBSubscription ensures the subscripion exists before listening for incoming messages.
+func (s *subscription) testSBSubscription(ctx context.Context) error {
+	sm, err := s.sbNs.NewSubscriptionManager(s.topicName)	
```

Comment:
You can test for the non-nilness of all the pieces you need here, instead of in `openSubscription`. That way `OpenSubscription` doesn't have to return an error.

Reply:
Make sense, fixed.